rfcs/20190802-model-garden-redesign.md

```diff
@@ -0,0 +1,207 @@
+# TensorFlow Official Model Garden Redesign
+
+| Status        | Proposed       |
+:-------------- |:---------------------------------------------------- |
+| **Author(s)** | Jing Li (jingli@google.com), Hongkun Yu (hongkuny@google.com), Xiaodan Song (xiaodansong@google.com) |
+| **Sponsor**   | Edd Wilder-James (ewj@google.com)                 |
+| **Updated**   | 2019-08-02                                           |
+
+## Objective
+
+This document presents a proposal to redesign TensorFlow official model garden.
+We aim to provide a central and reliable place to contain popular examples,
+state-of-the-art models and tutorials to demonstrate the best practice in TF2.0
+and illustrate real-world use cases.
+
+## Motivation
+
+The current [TF official model garden](https://github.com/tensorflow/models/tree/master/official)
+mainly has ad hoc support. Example models are implemented using mixed TensorFlow
+APIs in different coding styles and some of them have convergence and/or
+performance regression. With TensorFlow 2.0 launch, there’s a great desire to
+provide tensorflow users a clear and central place to showcase reliable TF2.0
+models with the best practices to follow.
+
+We want to take this opportunity to substantially improve the state of the
+official model garden, and provide seamlessly end-to-end training and inference
+user experience on a wide range of accelerators and mobile device chips. We hope
+to encourage community to contribute innovations and improve TensorFlow
+efficiency and usability.
+
+## User Benefit
+
+We aim to provide the best modeling experience via this revamp effort:
+
+*   Usability and reliability
+    *   keep official models well-maintained and tested for both performance and
+        convergence.
+    *   provide accessible model distribution via [TensorFlow Hub](https://www.tensorflow.org/hub) and share state-of-the-art research accomplishments.
+    *   make training on both GPU and TPU an easy switch.
+    *   provide reusable components for research and production.
+*   End-to-end solutions
+    *   provide seamless end-to-end training and inference solutions, where inference covers serving on TPU, GPU, mobile and edge devices.
+    *   provide hyper parameter sets to tune models for various resource constraints.
+    *   provide solutions with hyper parameters to scale model training to TPU pods or multi-worker GPUs.
+    *   provide variants derived from standard models to tackle various practical tasks.
+
+## Design Proposal
+
+### Official model directory reorgnization
+
+We are going to reorganize the official model directory to provide:
+
+*   common libraries, mainly two types:
```

Comment:
Moving common code into a central location is always a good design practice, but let's be careful not create another SLIM here. There are *still* models (Deeplab anyone) that are chained to SLIM.

Remember, part of what would be valuable here is showing good examples on how to write a *model*, not necessarily good examples on how to write a model *garden* with intertwined and complicated dependencies.

Reply:
I agree on the last part. `models/research` is a typical example of this. The whole garden there is so complex and huge that it is nowhere near usable for a project until unless you:
* Just want the model off the shelf without any modification
* You know in an out of each and every part of it

Reply:
Same concern. Software engineering design may not be applied to deep learning model garden these days.

Personally, a model *garden* is simply somewhere user can obtain **the whole runnable model** from a very sub directory. Sharing common library, and any similar things, sounds like a SDK built on top of TF to me. Maybe we can have TF SDK, and then the garden.

Reply:
Thank you for raising this concern! We definitely need a good balance between implementation duplication and complicated dependencies. The current design is to split the modeling process to two stages: 1)common networks (in modeling directory), such as resnet, transformer, so that these networks can be reused in specific model/task, e.g. resnet50, mask r-cnn, sequence model; 2) specific model/task along with public dataset (e.g. in NLP and vision directory), models will be defined in its own sub directory, data preprocessing and eval and other utils for the same type of tasks/datasets can be shared. 

If you have better suggestions, we are definitely open to them. Thanks again!

Reply:
> The current design is to split the modeling process to two stages: 1)common networks (in modeling directory), such as resnet, transformer, so that these networks can be reused in specific model/task, e.g. resnet50, mask r-cnn, sequence model

How will these be different from https://www.tensorflow.org/api_docs/python/tf/keras/applications ?

Reply:
Also, Keras applications are going on an external repo x Keras migration RFC at https://github.com/tensorflow/community/pull/202. 
Right?